Sources/TuistKit/Services/BuildService.swift

```diff
@@ -1,19 +1,83 @@
 import Foundation
+import RxBlocking
 import TSCBasic
+import TuistAutomation
+import TuistCore
 import TuistSupport
 
 enum BuildServiceError: FatalError {
+    case schemeNotFound(scheme: String, existing: [String])
+
     // Error description
     var description: String {
-        ""
+        switch self {
+        case let .schemeNotFound(scheme, existing):
+            return "Couldn't find scheme \(scheme). The available schemes are: \(existing.joined(separator: ", "))."
+        }
     }
 
     // Error type
-    var type: ErrorType { .abort }
+    var type: ErrorType {
+        switch self {
+        case .schemeNotFound:
+            return .abort
+        }
+    }
 }
 
 final class BuildService {
-    func run() throws {
-        logger.notice("Command not available yet")
+    /// Project generator
+    let projectGenerator: ProjectGenerating
+
+    /// Xcode build controller.
+    let xcodebuildController: XcodeBuildControlling
+
+    /// Build graph inspector.
+    let buildGraphInspector: BuildGraphInspecting
+
+    init(projectGenerator: ProjectGenerating = ProjectGenerator(),
+         xcodebuildController: XcodeBuildControlling = XcodeBuildController(),
+         buildGraphInspector: BuildGraphInspecting = BuildGraphInspector()) {
+        self.projectGenerator = projectGenerator
+        self.xcodebuildController = xcodebuildController
+        self.buildGraphInspector = buildGraphInspector
+    }
+
+    func run(schemeName: String?, generate: Bool, path: AbsolutePath) throws {
+        let graph: Graph
+        if generate || buildGraphInspector.workspacePath(directory: path) == nil {
+            graph = try projectGenerator.generateWithGraph(path: path, projectOnly: false).1
+        } else {
+            graph = try projectGenerator.load(path: path)
+        }
+
+        let buildableSchemes = buildGraphInspector.buildableSchemes(graph: graph)
+        logger.log(level: .notice, "Found the following buildable schemes: \(buildableSchemes.map(\.name).joined(separator: ", "))")
+        var cleaned: Bool = false
+
+        func buildScheme(scheme: Scheme) throws {
```

Comment:
I've never used nested functions like this ^_^
nit: Isn't it more readable to have a separate private func?

Reply:
I made it a private function in the class now. I agree with you that embedded functions are not very readable.